File: backend/src/modules/dispatcher/dispatcher.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, Like } from 'typeorm';
import { Dispatcher } from './entities/dispatcher.entity';
import { CreateDispatcherDto } from './dto/create-dispatcher.dto';
import { UpdateDispatcherDto } from './dto/update-dispatcher.dto';
import { PaginationDto, PaginatedResult } from '../../common/dto/pagination.dto';
import { KamailioRpcService } from '../../common/services/kamailio-rpc.service';

@Injectable()
export class DispatcherService {
  constructor(
    @InjectRepository(Dispatcher)
    private dispatcherRepository: Repository<Dispatcher>,
    private kamailioRpcService: KamailioRpcService,
  ) {}

  /**
   * 创建调度目标
   */
  async create(createDto: CreateDispatcherDto): Promise<Dispatcher> {
    const dispatcher = this.dispatcherRepository.create(createDto);
    const saved = await this.dispatcherRepository.save(dispatcher);

    // 通知 Kamailio 重载调度器
    try {
      await this.kamailioRpcService.reloadDispatcher();
    } catch (error) {
      // 忽略 RPC 错误
    }

    return saved;
  }

  /**
   * 分页查询调度目标
   */
  async findAll(paginationDto: PaginationDto): Promise<PaginatedResult<Dispatcher>> {
    const { page, limit, keyword } = paginationDto;
    const skip = (page - 1) * limit;

    const where: any = {};
    if (keyword) {
      where.destination = Like(`%${keyword}%`);
    }

    const [items, total] = await this.dispatcherRepository.findAndCount({
      where,
      skip,
      take: limit,
      order: { setid: 'ASC', priority: 'DESC', id: 'ASC' },
    });

    return new PaginatedResult(items, total, page, limit);
  }

  /**
   * 根据调度组获取调度目标
   */
  async findBySetId(setid: number): Promise<Dispatcher[]> {
    return this.dispatcherRepository.find({
      where: { setid },
      order: { priority: 'DESC', id: 'ASC' },
    });
  }

  /**
   * 获取所有调度组
   */
  async getSetIds(): Promise<number[]> {
    const result = await this.dispatcherRepository
      .createQueryBuilder('d')
      .select('DISTINCT d.setid', 'setid')
      .orderBy('d.setid', 'ASC')
      .getRawMany();

    return result.map((r) => r.setid);
  }

  /**
   * 根据 ID 查询调度目标
   */
  async findOne(id: number): Promise<Dispatcher> {
    const dispatcher = await this.dispatcherRepository.findOne({ where: { id } });
    if (!dispatcher) {
      throw new NotFoundException('调度目标不存在');
    }
    return dispatcher;
  }

  /**
   * 更新调度目标
   */
  async update(id: number, updateDto: UpdateDispatcherDto): Promise<Dispatcher> {
    const dispatcher = await this.findOne(id);
    
    Object.assign(dispatcher, updateDto);

    const saved = await this.dispatcherRepository.save(dispatcher);

    // 通知 Kamailio 重载调度器
    try {
      await this.kamailioRpcService.reloadDispatcher();
    } catch (error) {
      // 忽略 RPC 错误
    }

    return saved;
  }

  /**
   * 删除调度目标
   */
  async remove(id: number): Promise<void> {
    const dispatcher = await this.findOne(id);
    await this.dispatcherRepository.remove(dispatcher);

    // 通知 Kamailio 重载调度器
    try {
      await this.kamailioRpcService.reloadDispatcher();
    } catch (error) {
      // 忽略 RPC 错误
    }
  }

  /**
   * 重载调度器配置
   */
  async reload(): Promise<void> {
    await this.kamailioRpcService.reloadDispatcher();
  }

  /**
   * 获取调度器实时状态
   */
  async getStatus(): Promise<any> {
    return this.kamailioRpcService.getDispatcherList();
  }

  /**
   * 获取统计信息
   */
  async getStats(): Promise<{ total: number; groups: { setid: number; count: number }[] }> {
    const total = await this.dispatcherRepository.count();
    
    const groups = await this.dispatcherRepository
      .createQueryBuilder('d')
      .select('d.setid', 'setid')
      .addSelect('COUNT(*)', 'count')
      .groupBy('d.setid')
      .orderBy('d.setid', 'ASC')
      .getRawMany();

    return { total, groups };
  }
}

File: kam-admin-module-routing/src/main/java/io/kamailio/admin/modules/dispatcher/service/DispatcherService.java
package io.kamailio.admin.modules.dispatcher.service;

import io.kamailio.admin.common.RpcTimeouts;
import io.kamailio.admin.common.dto.PaginatedResult;
import io.kamailio.admin.common.dto.PaginationDto;
import io.kamailio.admin.common.service.KamailioRpcService;
import io.kamailio.admin.modules.dispatcher.dto.CreateDispatcherDto;
import io.kamailio.admin.modules.dispatcher.dto.UpdateDispatcherDto;
import io.kamailio.admin.modules.dispatcher.entity.Dispatcher;
import io.kamailio.admin.modules.dispatcher.repository.DispatcherRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Map;
@Service
public class DispatcherService {

    private final DispatcherRepository repository;
    private final KamailioRpcService kamailioRpc;

    public DispatcherService(DispatcherRepository repository, KamailioRpcService kamailioRpc) {
        this.repository = repository;
        this.kamailioRpc = kamailioRpc;
    }

    private void reloadDispatcher() {
        kamailioRpc.reloadDispatcher().subscribe(
                v -> {},
                e -> {}
        );
    }

    @Transactional
    public Dispatcher create(CreateDispatcherDto dto) {
        var entity = new Dispatcher();
        entity.setSetid(dto.getSetid());
        entity.setDestination(dto.getDestination());
        entity.setFlags(dto.flagsOrDefault());
        entity.setPriority(dto.priorityOrDefault());
        entity.setAttrs(dto.attrsOrDefault());
        entity.setDescription(dto.descriptionOrDefault());
        var saved = repository.save(entity);
        reloadDispatcher();
        return saved;
    }

    public PaginatedResult<Dispatcher> findAll(PaginationDto dto) {
        Page<Dispatcher> page;
        if (dto.getKeyword() != null && !dto.getKeyword().isBlank()) {
            page = repository.findByDestinationContaining(dto.getKeyword(),
                    PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(),
                            Sort.by(Sort.Direction.ASC, "setid")
                                    .and(Sort.by(Sort.Direction.DESC, "priority"))
                                    .and(Sort.by(Sort.Direction.ASC, "id"))));
        } else {
            page = repository.findAll(PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(),
                    Sort.by(Sort.Direction.ASC, "setid")
                            .and(Sort.by(Sort.Direction.DESC, "priority"))
                            .and(Sort.by(Sort.Direction.ASC, "id"))));
        }
        return new PaginatedResult<>(page.getContent(), page.getTotalElements(),
                dto.pageOrDefault(), dto.limitOrDefault());
    }

    public List<Dispatcher> findBySetId(Integer setid) {
        return repository.findBySetidOrderByPriorityDescIdAsc(setid);
    }

    public List<Integer> getSetIds() {
        return repository.findDistinctSetIds();
    }

    public Dispatcher findOne(Integer id) {
        return repository.findById(id)
                .orElseThrow(() -> new RuntimeException("调度目标不存在"));
    }

    @Transactional
    public Dispatcher update(Integer id, UpdateDispatcherDto dto) {
        var entity = findOne(id);
        if (dto.getSetid() != null) entity.setSetid(dto.getSetid());
        if (dto.getDestination() != null) entity.setDestination(dto.getDestination());
        if (dto.getFlags() != null) entity.setFlags(dto.getFlags());
        if (dto.getPriority() != null) entity.setPriority(dto.getPriority());
        if (dto.getAttrs() != null) entity.setAttrs(dto.getAttrs());
        if (dto.getDescription() != null) entity.setDescription(dto.getDescription());
        var saved = repository.save(entity);
        reloadDispatcher();
        return saved;
    }

    @Transactional
    public void remove(Integer id) {
        repository.delete(findOne(id));
        reloadDispatcher();
    }

    public void reload() {
        kamailioRpc.reloadDispatcher().block(RpcTimeouts.DEFAULT_BLOCK);
    }

    public Object getStatus() {
        return kamailioRpc.getDispatcherList().block(RpcTimeouts.DEFAULT_BLOCK);
    }

    public void setState(String state, Integer group, String address) {
        kamailioRpc.call("dispatcher.set_state", state, group, address).block(RpcTimeouts.DEFAULT_BLOCK);
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> getStats() {
        var total = repository.count();
        var rows = repository.countBySetId();
        var groups = rows.stream()
                .map(r -> Map.<String, Object>of("setid", r[0], "count", r[1]))
                .toList();
        return Map.of("total", total, "groups", groups);
    }
}
